pkg/ccl/streamingccl/streamingest/stream_ingestion_processor_test.go

```diff
@@ -171,6 +203,61 @@ func TestStreamIngestionProcessor(t *testing.T) {
 		require.Nil(t, row)
 		testutils.IsError(meta.Err, "this client always returns an error")
 	})
+
+	t.Run("stream ingestion processor shuts down gracefully on losing client connection", func(t *testing.T) {
+		events := []streamingccl.Event{streamingccl.MakeGenerationEvent()}
+		pa := streamingccl.PartitionAddress("partition")
+		mockClient := &mockStreamClient{
+			partitionEvents: map[streamingccl.PartitionAddress][]streamingccl.Event{pa: events},
+		}
+
+		startTime := hlc.Timestamp{WallTime: timeutil.Now().UnixNano()}
+		partitionAddresses := []streamingccl.PartitionAddress{"partition"}
+
+		interceptCh := make(chan struct{})
+		defer close(interceptCh)
+		sendToInterceptCh := func() {
+			interceptCh <- struct{}{}
+		}
+		interceptGeneration := markGenerationEventReceived(sendToInterceptCh)
+		sip, out, err := getStreamIngestionProcessor(ctx, t, registry, kvDB, "randomgen://test/",
+			partitionAddresses, startTime, []streamclient.InterceptFn{interceptGeneration}, mockClient)
+		require.NoError(t, err)
+
+		var wg sync.WaitGroup
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			sip.Run(ctx)
+		}()
+
+		// The channel will block on read if the event has not been intercepted yet.
+		// Once it unblocks, we are guaranteed that the mockClient has sent the
+		// GenerationEvent and the processor has read it.
+		<-interceptCh
+
+		// The sip processor has received a GenerationEvent and is thus
+		// waiting for a cutover signal, so let's send one!
+		sip.cutoverCh <- struct{}{}
+
+		wg.Wait()
+		// Ensure that all the outputs are properly closed.
+		if !out.ProducerClosed() {
+			t.Fatalf("output RowReceiver not closed")
+		}
+
+		for {
+			row := out.NextNoMeta(t)
+			if row == nil {
+				break
+			}
+			t.Fatalf("more output rows than expected")
+		}
+
+		// The processor should have been moved to draining state with a nil error
+		_, meta := sip.Next()
```

Comment:
I don't think we need this any longer. `out` is a row buffer that is filled by `sip.Next()`. The `out.NextNoMeta` checks that none of the output was a meta.